vendor/bat-native-ledger/src/bat/ledger/internal/response/response_promotion.cc

```diff
@@ -0,0 +1,306 @@
+//       "approximateValue": "30",
+//       "type": "ugp",
+//       "available": true,
+//       "platform": "desktop",
+//       "publicKeys": [
+//         "dvpysTSiJdZUPihius7pvGOfngRWfDiIbrowykgMi1I="
+//       ],
+//       "legacyClaimed": false
+//     }
+//   ]
+// }
+
+ledger::Result CheckFetchPromotions(const ledger::UrlResponse& response) {
+  // Bad Request (400)
+  if (response.status_code == net::HTTP_BAD_REQUEST) {
+    BLOG(0, "Invalid paymentId or platform in request");
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  // Not Found (404)
+  if (response.status_code == net::HTTP_NOT_FOUND) {
+    BLOG(0, "Unrecognized paymentId/promotion combination");
+    return ledger::Result::NOT_FOUND;
+  }
+
+  // Internal Server Error (500)
+  if (response.status_code == net::HTTP_INTERNAL_SERVER_ERROR) {
+    BLOG(0, "Internal server error");
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  if (response.status_code != net::HTTP_OK) {
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  return ledger::Result::LEDGER_OK;
+}
+
+ledger::Result ParseFetchPromotions(
+    const ledger::UrlResponse& response,
+    ledger::PromotionList* list,
+    std::vector<std::string>* corrupted_promotions) {
+  DCHECK(list && corrupted_promotions);
+
+  base::Optional<base::Value> value = base::JSONReader::Read(response.body);
+  if (!value || !value->is_dict()) {
+    BLOG(0, "Invalid JSON");
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  base::DictionaryValue* dictionary = nullptr;
+  if (!value->GetAsDictionary(&dictionary)) {
+    BLOG(0, "Invalid JSON");
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  auto* promotions = dictionary->FindListKey("promotions");
+  if (!promotions) {
+    return ledger::Result::LEDGER_OK;
+  }
+
+  const auto promotion_size = promotions->GetList().size();
+  for (auto& item : promotions->GetList()) {
+    ledger::PromotionPtr promotion = ledger::Promotion::New();
+
+    const auto* id = item.FindStringKey("id");
+    if (!id) {
+      continue;
+    }
+    promotion->id = *id;
+
+    const auto version = item.FindIntKey("version");
+    if (!version) {
+      corrupted_promotions->push_back(promotion->id);
+      continue;
+    }
+    promotion->version = *version;
+
+    const auto* type = item.FindStringKey("type");
+    if (!type) {
+      corrupted_promotions->push_back(promotion->id);
+      continue;
+    }
+    promotion->type =
+        braveledger_promotion::ConvertStringToPromotionType(*type);
+
+    const auto suggestions = item.FindIntKey("suggestionsPerGrant");
+    if (!suggestions) {
+      corrupted_promotions->push_back(promotion->id);
+      continue;
+    }
+    promotion->suggestions = *suggestions;
+
+    const auto* approximate_value = item.FindStringKey("approximateValue");
+    if (!approximate_value) {
+      corrupted_promotions->push_back(promotion->id);
+      continue;
+    }
+    base::StringToDouble(*approximate_value, &promotion->approximate_value);
```

Comment:
we need to handle return so code would be something like this

```suggestion
   const bool success = base::StringToDouble(*approximate_value, &promotion->approximate_value);
    if (!success) {
      promotion->approximate_value = 0.0;
    }
```

Reply:
Fixed.